File: code/programs/elixir/unix-tools/test/dirname_test.exs
defmodule DirnameTest do
  @moduledoc """
  Tests for the dirname tool.

  ## What These Tests Verify

  These tests cover:
  1. CLI Builder integration (parsing flags, --help, --version, errors).
  2. Business logic (compute_dirname with various paths).
  3. Edge cases (trailing slashes, root, no slashes, empty string).
  """

  use ExUnit.Case, async: true

  alias CodingAdventures.CliBuilder.Parser
  alias CodingAdventures.CliBuilder.{ParseResult, HelpResult, VersionResult, ParseErrors}

  # ---------------------------------------------------------------------------
  # Helper: locate the spec file
  # ---------------------------------------------------------------------------

  @spec_path Path.join([__DIR__, "..", "dirname.json"]) |> Path.expand()

  # ---------------------------------------------------------------------------
  # Helper: parse argv against the dirname spec
  # ---------------------------------------------------------------------------

  defp parse_argv(argv) do
    Parser.parse(@spec_path, argv)
  end

  # ---------------------------------------------------------------------------
  # Test: Default behavior
  # ---------------------------------------------------------------------------

  describe "default behavior" do
    test "single name argument is captured" do
      {:ok, %ParseResult{arguments: arguments}} = parse_argv(["dirname", "/usr/bin"])
      assert arguments["names"] == ["/usr/bin"]
    end

    test "multiple name arguments are captured" do
      {:ok, %ParseResult{arguments: arguments}} =
        parse_argv(["dirname", "/usr/bin", "/etc/hosts"])

      assert arguments["names"] == ["/usr/bin", "/etc/hosts"]
    end
  end

  # ---------------------------------------------------------------------------
  # Test: Flags
  # ---------------------------------------------------------------------------

  describe "flags" do
    test "-z sets zero to true" do
      {:ok, %ParseResult{flags: flags}} = parse_argv(["dirname", "-z", "/usr/bin"])
      assert flags["zero"] == true
    end
  end

  # ---------------------------------------------------------------------------
  # Test: --help and --version
  # ---------------------------------------------------------------------------

  describe "--help flag" do
    test "returns {:ok, %HelpResult{}}" do
      assert {:ok, %HelpResult{}} = parse_argv(["dirname", "--help"])
    end

    test "help text contains program name" do
      {:ok, %HelpResult{text: text}} = parse_argv(["dirname", "--help"])
      assert text =~ "dirname"
    end
  end

  describe "--version flag" do
    test "returns {:ok, %VersionResult{}}" do
      assert {:ok, %VersionResult{}} = parse_argv(["dirname", "--version"])
    end

    test "version string is 1.0.0" do
      {:ok, %VersionResult{version: version}} = parse_argv(["dirname", "--version"])
      assert version == "1.0.0"
    end
  end

  # ---------------------------------------------------------------------------
  # Test: Unknown flags
  # ---------------------------------------------------------------------------

  describe "unknown flags" do
    test "unknown long flag returns error" do
      assert {:error, %ParseErrors{}} = parse_argv(["dirname", "--unknown", "/usr"])
    end
  end

  # ---------------------------------------------------------------------------
  # Test: Business logic - compute_dirname
  # ---------------------------------------------------------------------------

  describe "compute_dirname/1" do
    test "normal absolute path" do
      assert UnixTools.DirnameTool.compute_dirname("/usr/bin") == "/usr"
    end

    test "path with trailing slash" do
      assert UnixTools.DirnameTool.compute_dirname("/usr/") == "/"
    end

    test "path without slashes returns dot" do
      assert UnixTools.DirnameTool.compute_dirname("usr") == "."
    end

    test "root path returns root" do
      assert UnixTools.DirnameTool.compute_dirname("/") == "/"
    end

    test "dot returns dot" do
      assert UnixTools.DirnameTool.compute_dirname(".") == "."
    end

    test "double dot returns dot" do
      assert UnixTools.DirnameTool.compute_dirname("..") == "."
    end

    test "empty string returns dot" do
      assert UnixTools.DirnameTool.compute_dirname("") == "."
    end

    test "deep path" do
      assert UnixTools.DirnameTool.compute_dirname("/usr/local/bin/node") == "/usr/local/bin"
    end

    test "double-slash path" do
      assert UnixTools.DirnameTool.compute_dirname("//usr") == "/"
    end

    test "relative path with directory" do
      assert UnixTools.DirnameTool.compute_dirname("include/stdio.h") == "include"
    end
  end
end

File: code/programs/rust/unix-tools/tests/dirname_test.rs
//! # Integration Tests for dirname
//!
//! These tests exercise both the CLI Builder integration and the dirname
//! business logic. We test directory extraction for various path formats
//! including absolute, relative, root, and edge cases.

use cli_builder::{load_spec_from_file, Parser, ParserOutput};
use unix_tools::dirname_tool::strip_dirname;
use std::path::PathBuf;

// ---------------------------------------------------------------------------
// Helper: locate the spec file
// ---------------------------------------------------------------------------

fn spec_path() -> String {
    let manifest_dir = env!("CARGO_MANIFEST_DIR");
    let path = PathBuf::from(manifest_dir).join("dirname.json");
    path.to_string_lossy().into_owned()
}

fn make_parser() -> Parser {
    let spec = load_spec_from_file(&spec_path()).expect("failed to load dirname.json");
    Parser::new(spec)
}

fn parse_argv(argv: &[&str]) -> ParserOutput {
    let parser = make_parser();
    let args: Vec<String> = argv.iter().map(|s| s.to_string()).collect();
    parser.parse(&args).expect("parse failed")
}

// ---------------------------------------------------------------------------
// Test: Spec loads successfully
// ---------------------------------------------------------------------------

#[cfg(test)]
mod spec_loading {
    use super::*;

    #[test]
    fn spec_loads() {
        let spec = load_spec_from_file(&spec_path());
        assert!(spec.is_ok(), "dirname.json should load successfully");
    }
}

// ---------------------------------------------------------------------------
// Test: Flag parsing
// ---------------------------------------------------------------------------

#[cfg(test)]
mod flag_parsing {
    use super::*;

    #[test]
    fn with_name_arg() {
        match parse_argv(&["dirname", "/usr/bin/sort"]) {
            ParserOutput::Parse(_) => {}
            other => panic!("expected Parse, got {:?}", other),
        }
    }

    #[test]
    fn zero_flag() {
        match parse_argv(&["dirname", "-z", "/usr/bin"]) {
            ParserOutput::Parse(result) => {
                assert_eq!(
                    result.flags.get("zero"),
                    Some(&serde_json::json!(true))
                );
            }
            _ => panic!("expected Parse"),
        }
    }

    #[test]
    fn help_returns_help() {
        match parse_argv(&["dirname", "--help"]) {
            ParserOutput::Help(_) => {}
            other => panic!("expected Help, got {:?}", other),
        }
    }

    #[test]
    fn version_returns_version() {
        match parse_argv(&["dirname", "--version"]) {
            ParserOutput::Version(v) => {
                assert_eq!(v.version, "1.0.0");
            }
            _ => panic!("expected Version"),
        }
    }
}

// ---------------------------------------------------------------------------
// Test: Business logic — strip_dirname
// ---------------------------------------------------------------------------

#[cfg(test)]
mod business_logic {
    use super::*;

    #[test]
    fn simple_path() {
        assert_eq!(strip_dirname("/usr/bin/sort"), "/usr/bin");
    }

    #[test]
    fn trailing_slash() {
        assert_eq!(strip_dirname("/usr/bin/"), "/usr");
    }

    #[test]
    fn bare_filename() {
        assert_eq!(strip_dirname("hello"), ".");
    }

    #[test]
    fn root_path() {
        assert_eq!(strip_dirname("/"), "/");
    }

    #[test]
    fn empty_string() {
        assert_eq!(strip_dirname(""), ".");
    }

    #[test]
    fn relative_path() {
        assert_eq!(strip_dirname("a/b"), "a");
    }

    #[test]
    fn file_in_root() {
        assert_eq!(strip_dirname("/hello"), "/");
    }

    #[test]
    fn deep_path() {
        assert_eq!(strip_dirname("/a/b/c/d"), "/a/b/c");
    }
}
